File: kotlin/src/main/kotlin/tasks/miniaturesUpdate/ScritchMiniatureUrlResolver.kt
package tasks.miniaturesUpdate

import data.JsonException
import data.JsonNavigator
import io.ktor.http.*
import web.client.CookieEagerHttpClient
import web.client.FastHttpClient
import web.client.GentleHttpClient
import web.client.HttpClientInterface
import web.snapshots.Snapshot
import web.url.FreeUrl
import web.url.Url

class ScritchMiniatureUrlResolver(
    httpClient: HttpClientInterface? = null,
) : MiniatureUrlResolver {
    private val graphQlUrl = FreeUrl("https://scritch.es/graphql")

    private val httpClient = httpClient ?: CookieEagerHttpClient(GentleHttpClient(FastHttpClient()))
    private val regex: Regex = Regex("^https://scritch\\.es/pictures/(?<pictureId>[-a-f0-9]{36})\$")

    override fun supports(url: String) = regex.containsMatchIn(url)

    override fun getMiniatureUrl(url: Url): String {
        val pictureId = regex.find(url.getUrl())?.groups?.get("pictureId")?.value
            ?: throw MiniatureUrlResolverException("Failed matching pictureId in the photo URL")

        val response = getResponseForPictureId(pictureId)

        if (response.metadata.httpCode != 200) {
            throw MiniatureUrlResolverException("Non-200 HTTP response code")
        }

        try {
            return JsonNavigator(response.contents).getNonEmptyString("data/medium/thumbnail")
        } catch (exception: JsonException) {
            throw MiniatureUrlResolverException("Wrong JSON data", exception)
        }
    }

    private fun getResponseForPictureId(pictureId: String): Snapshot {
        val csrfToken = getCsrfToken()
        val jsonPayload = getGraphQlJsonPayload(pictureId)

        val headers = mapOf(
            "Content-Type" to "application/json",
            "X-CSRF-Token" to csrfToken,
            "authorization" to "Scritcher $csrfToken",
        )

        return httpClient.fetch(graphQlUrl, HttpMethod.Post, headers, jsonPayload)
    }

    private fun getGraphQlJsonPayload(pictureId: String): String {
        return "{\"operationName\": \"Medium\", \"variables\": {\"id\": \"$pictureId\"}, \"query\": \"query " +
                "Medium(\$id: ID!, \$tagging: Boolean) { medium(id: \$id, tagging: \$tagging) { thumbnail } }\"}"
    }

    private fun getCsrfToken(): String = getOptionalCsrfToken() ?: getFirstRequiredCsrfToken()

    private fun getOptionalCsrfToken(): String? {
        return httpClient.getSingleCookieValue("https://scritch.es/", "csrf-token")
    }

    private fun getFirstRequiredCsrfToken(): String {
        httpClient.fetch(FreeUrl("https://scritch.es/"))

        return getOptionalCsrfToken() ?: throw MiniatureUrlResolverException("Missing csrf-token cookie")
    }
}

File: symfony/src/Photos/MiniaturesFinder/ScritchMiniatureUrlResolver.php
<?php

declare(strict_types=1);

namespace App\Photos\MiniaturesFinder;

use App\Utils\Web\FreeUrl;
use App\Utils\Web\HttpClient\HttpClientInterface;
use App\Utils\Web\Snapshots\Snapshot;
use App\Utils\Web\Url;
use Override;
use TRegx\CleanRegex\Pattern;
use Veelkoov\Debris\StringStringMap;

class ScritchMiniatureUrlResolver implements MiniatureUrlResolver
{
    private readonly Pattern $pattern;

    public function __construct(
        private readonly HttpClientInterface $httpClient,
    ) {
        $this->pattern = Pattern::of('^https://scritch\.es/pictures/(?<pictureId>[-a-f0-9]{36})$');
    }

    #[Override]
    public function supports(string $url): bool
    {
        return $this->pattern->test($url);
    }

    #[Override]
    public function getMiniatureUrl(Url $url): string
    {
        $pictureId = $this->pattern->match($url->getUrl())->first()->get('pictureId');

        $response = $this->getResponseForPictureId($pictureId);

        //        if (response.metadata.httpCode != 200) {
        //            throw MiniatureUrlResolverException("Non-200 HTTP response code")
        //        }
        //
        //        try {
        //            return JsonNavigator(response.contents).getNonEmptyString("data/medium/thumbnail")
        //        } catch (exception: JsonException) {
        //            throw MiniatureUrlResolverException("Wrong JSON data", exception)
        //        }
        return ''; // FIXME
    }

    private function getResponseForPictureId(string $pictureId): Snapshot
    {
        $csrfToken = $this->getCsrfToken();
        $jsonPayload = $this->getGraphQlJsonPayload($pictureId);

        $headers = new StringStringMap([
            'Content-Type' => 'application/json',
            'X-CSRF-Token' => $csrfToken,
            'authorization' => "Scritcher $csrfToken",
        ]);

        return $this->httpClient->fetch(new FreeUrl('https://scritch.es/graphql'), 'POST', $headers, $jsonPayload);
    }

    private function getGraphQlJsonPayload(string $pictureId): string
    {
        return <<<GRAPHQL
            {   
                "operationName": "Medium",
                "variables": {"id": "$pictureId"},
                "query": "query Medium(\$id: ID!, \$tagging: Boolean) {
                    medium(id: \$id, tagging: \$tagging) { thumbnail } 
                }"
            }
        GRAPHQL;
    }

    private function getCsrfToken(): string
    {
        return $this->getOptionalCsrfToken() ?? $this->getFirstRequiredCsrfToken();
    }

    private function getOptionalCsrfToken(): ?string
    {
        return $this->httpClient->getSingleCookieValue('https://scritch.es/', 'csrf-token');
    }

    private function getFirstRequiredCsrfToken(): string
    {
        $this->httpClient->fetch(new FreeUrl('https://scritch.es/'));

        return $this->getOptionalCsrfToken() ?? throw new MiniatureFinderException('Missing csrf-token cookie.');
    }
}
